lecture2gether_flask/meta_data_provider.py

```diff
@@ -150,3 +157,55 @@ def youtube_video_id_from_url(video_url):
     return None
 
 
+class GoogleDriveMetaDataProvider(MetaDataProvider):
+    def __init__(self, share_link):
+        super().__init__(share_link)
+        self._file_id = drive_file_id_from_share_url(share_link)
+
+        if not self._file_id:
+            raise VideoNotFoundException
+
+        # Check API Key
+        if 'GOOGLE_API_KEY' not in os.environ:
+            raise APIUnauthorizedException
+
+        # Init Google API
+        try:
+            self._drive = googleapiclient.discovery.build(
+                'drive', 'v3', developerKey=os.environ['GOOGLE_API_KEY'], cache_discovery=False)
+        except HttpError as e:
+            if e.resp.status in [400, 401, 403]:
+                raise APIUnauthorizedException
+            else:
+                raise e
+
+    def get_meta_data(self):
+        # Get meta data for file
+        try:
+            file_meta = self._drive.files().get(fileId=self._file_id).execute()
+        except HttpError as e:
+            if e.resp.status in [400, 401, 403]:
+                raise VideoUnauthorizedException
+            elif e.resp.status in [404,]:
+                raise VideoNotFoundException
+            else:
+                raise e
+        self.video_meta_data["title"] = file_meta["name"]
+        self.video_meta_data["mimeType"] = file_meta["mimeType"]
+
+        # Get stream url from API
+        file_download_url = json.loads(self._drive.files().get_media(fileId=self._file_id).to_json())['uri']
```

Comment:
Why do you first put this into json to then extract one key?

Reply:
Because that is the seemd like the best way to access this key. Or is there another method to access the requests uri without sending the request?

Reply:
What does `self._drive.files().get_media(fileId=self._file_id)` return and what data structure?
I just don't understand, why to convert the output to json and then accessing the uri instead of accessing the uri from the output directly.

Reply:
It returns a Google API request object. Normally this object would call the API by calling `.execute()` on it. But we want only the API call to send it to the frontend. Therefore we use the `.to_json()` call to get the request as a string containing a bit of JSON. I then parse this JSON to access the URI without reading private fields or stuff like that.